lerobot/common/policies/vqbet/modeling_vqbet.py

```diff
@@ -0,0 +1,2460 @@
+
+        self.register_buffer('discretized', torch.tensor(False))
+        self.optimized_steps = 0
+
+        self.vq_layer = ResidualVQ(
+            dim=config.vqvae_embedding_dim,
+            num_quantizers=config.vqvae_groups,
+            codebook_size=config.vqvae_n_embed,
+        )
+
+        self.encoder = MLP(
+            in_channels=self.config.output_shapes["action"][0] * self.config.action_chunk_size,
+            hidden_channels=[config.vqvae_enc_hidden_dim, config.vqvae_enc_hidden_dim, config.vqvae_embedding_dim],
+        )
+        self.decoder = MLP(
+            in_channels=config.vqvae_embedding_dim,
+            hidden_channels=[config.vqvae_enc_hidden_dim, config.vqvae_enc_hidden_dim, self.config.output_shapes["action"][0] * self.config.action_chunk_size],
+        )
+
+        self.train()
+
+    def toggle_discretized(self, state=True):
+        self.discretized = torch.tensor(state)
+
+    def check_discretized(self):
+        return self.discretized.item()
+
+    def eval(self):
+        self.training = False
+        self.vq_layer.eval()
+        self.encoder.eval()
+        self.decoder.eval()
+
+    def train(self, mode=True):
+        """
+        This function forces the RVQ to no longer update when action discretization is complete.
+        Since VQs are partly updated via the EMA method, simply passing data through them can cause unintended modifications.
+        Therefore, we use function overriding to prevent RVQs from being updated during the training of VQ-BeT after discretization completes.
+        """
+        if mode:
+            if self.check_discretized():
+                pass
+            else:
+                self.training = True
+                self.vq_layer.train()
+                self.decoder.train()
+                self.encoder.train()
+        else:
+            self.eval()
+
+    def draw_logits_forward(self, encoding_logits):
+        z_embed = self.vq_layer.draw_logits_forward(encoding_logits)
+        return z_embed
+
+    def draw_code_forward(self, encoding_indices):
+        with torch.no_grad():
+            z_embed = self.vq_layer.get_codes_from_indices(encoding_indices)
+            z_embed = z_embed.sum(dim=0)
+        return z_embed
+
+    def get_action_from_latent(self, latent):
+        output = self.decoder(latent)
+        if self.config.action_chunk_size == 1:
+            return einops.rearrange(output, "N (T A) -> N T A", A=self.config.output_shapes["action"][0])
+        else:
+            return einops.rearrange(output, "N (T A) -> N T A", A=self.config.output_shapes["action"][0])
+
+    def preprocess(self, state):
+        if not torch.is_tensor(state):
+            state = torch.FloatTensor(state.copy())
+        if self.config.action_chunk_size == 1:
+            state = state.squeeze(-2)
+        else:
+            state = einops.rearrange(state, "N T A -> N (T A)")
+        return state
+
+    def get_code(self, state, required_recon=False):
```

Comment:
This looks like it's never called with `required_recon=True` and there appears to be bout 10 lines of logic we might be able to do away with.

On another note, can you please take a look through the rest of the code and find any code that is not accessed? I think we should make a determination about whether it is relevant (now or in the _near_ future) or not, and if not, remove it. It makes the code easier to read and maintain.

Reply:
Removed the unused parts. Thank you:)